packages/swarm-components/src/Dropdown.jsx

```diff
@@ -0,0 +1,453 @@
+					} else if (event.key === 'ArrowUp') {
+						event.preventDefault(); // prevent scroll
+						setState({ ...state, ...openAtFirstItem() });
+					}
+				})}
+				{...props}
+			/>
+		);
+	}
+);
+
+MenuButton.propTypes = {
+	onClick: func,
+	onKeyDown: func,
+	children: node,
+};
+
+const MenuItem = React.forwardRef(
+	(
+		{
+			onSelect,
+			onClick,
+			role = 'menuitem',
+			index,
+			onKeyDown,
+			onMouseMove,
+			onMouseLeave,
+			setState: propSetState,
+			state: propState,
+			_ref,
+			...rest
+		},
+		ref
+	) => {
+		const { state, setState } = useContext(MenuContext);
+		const isSelected = state && index === state.selectionIndex;
+		const select = () => {
+			onSelect();
+			setState({ ...state, ...closeState() });
+		};
+
+		const itemRef = useRef(null);
+
+		useEffect(() => {
+			if (itemRef.current && state.selectionIndex === index) {
+				itemRef.current.focus();
+			} else {
+				itemRef.current.blur();
+			}
+		}, [state.selectionIndex]);
+		return (
+			<div
+				{...rest}
+				ref={itemRef}
+				data-reach-menu-item={role === 'menuitem' ? true : undefined}
+				role={role}
+				tabIndex="-1"
+				data-selected={role === 'menuitem' && isSelected ? true : undefined}
+				onClick={ReachUtils.wrapEvent(onClick, event => {
+					select();
+				})}
+				onKeyDown={ReachUtils.wrapEvent(onKeyDown, event => {
+					if (event.key === 'Enter') {
+						// prevent the button from being "clicked" by
+						// this "Enter" keydown
+						event.preventDefault();
+						select();
+					}
+				})}
+				onMouseMove={ReachUtils.wrapEvent(onMouseMove, event => {
+					if (!isSelected) {
+						setState({ ...state, selectionIndex: index });
+					}
+				})}
+				onMouseLeave={ReachUtils.wrapEvent(onMouseLeave, event => {
+					// clear out selection when mouse over a non-menu item child
+					setState({ ...state, selectionIndex: -1 });
+				})}
+			/>
+		);
+	}
+);
+
+MenuItem.propTypes = {
+	onSelect: func.isRequired,
+	onClick: func,
+	role: string,
+	state: object,
+	setState: func,
+	index: number,
+	onKeyDown: func,
+	onMouseMove: func,
+	_ref: func,
+};
+
+const k = () => {};
+
+////////////////////////////////////////////////////////////////////////
+const MenuLink = React.forwardRef(
+	(
+		{
+			onKeyDown,
+			onClick,
+			component: Comp,
+			as: AsComp = 'a',
+			style,
+			setState,
+			state,
+			index,
+			_ref,
+			...props
+		},
+		ref
+	) => {
+		const Link = Comp || AsComp;
+		if (Comp) {
+			console.warn(
+				'[@swarm-ui/dropdown]: Please use the `as` prop instead of `component`.'
+			);
+		}
+		return (
+			<MenuItem role="none" index={index} onSelect={k} _ref={k}>
+				<Link
+					role="menuitem"
+					data-reach-menu-item
+					tabIndex="-1"
+					data-selected={index === state.selectionIndex ? true : undefined}
+					onClick={ReachUtils.wrapEvent(onClick, event => {
+						setState({ ...state, ...closeState() });
+					})}
+					onKeyDown={ReachUtils.wrapEvent(onKeyDown, event => {
```

Comment:
for anyone who wants to look at some of the reach utils, https://github.com/reach/reach-ui/blob/master/packages/utils/src/index.js#L37

i'm glad we're using these instead of making our own version